File: kangaroo_app_sdk_ios/ios/Classes/features/userPinResetRequest/UserPinResetRequestHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class UserPinResetRequestHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/request_pin_reset"

    var eventChannel: String = "customer_sdk/events/request_pin_reset"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await UserPinResetRequestHandler.requestPinReset(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func requestPinReset(call: FlutterMethodCall) async -> String? {
        

        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
                        guard let mode = myArgs["mode"] as? String else {return nil}
                let email = myArgs["email"] as? String? ?? nil
                let phone = myArgs["phone"] as? String? ?? nil
                let countryCode = myArgs["countryCode"] as? String? ?? nil

         let result = try await UserPinResetRequestApi().requestPinReset(
                mode: mode,
                email: email,
                phone: phone,
                countryCode: countryCode
            ).serializeNative()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        UserPinResetRequestApi().userPinResetRequestState
            .serializeUserPinResetRequestState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}

File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/userPinResetRequest/UserPinResetRequestHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.userPinResetRequest

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.userPinResetRequest.UserPinResetRequestApi
import features.userPinResetRequest.models.UserPinResetRequest

import features.userPinResetRequest.serializeUserPinResetRequestState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class UserPinResetRequestHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/request_pin_reset"

    override val eventChannel: String
        get() = "customer_sdk/events/request_pin_reset"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return requestPinReset(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun requestPinReset(call: MethodCall): String? {
            val result = UserPinResetRequestApi().requestPinReset(
                mode = call.argument<String>("mode") as String,
                email = call.argument<String?>("email"),
                phone = call.argument<String?>("phone"),
                countryCode = call.argument<String?>("countryCode")
            ).toJsonResult<UserPinResetRequest>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        UserPinResetRequestApi().userPinResetRequestState.serializeUserPinResetRequestState().observe { result ->
            pushSerializedResultToEventSink<UserPinResetRequestApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}
